axelor-contract/src/main/java/com/axelor/apps/contract/service/ContractLineServiceImpl.java

```diff
@@ -213,4 +227,45 @@ public void updateContractLinesFromContractVersion(ContractVersion contractVersi
     }
     contractVersionRepo.save(contractVersion);
   }
+
+  @Override
+  public ContractLine computePricesPerYear(
+      ContractLine contractLine, ContractVersion contractVersion) throws AxelorException {
+    Duration duration = contractVersion.getInvoicingDuration();
+
+    if (duration != null) {
+      BigDecimal initialUnitPrice = contractLine.getInitialUnitPrice();
+      BigDecimal qty = contractLine.getQty();
+      BigDecimal exTaxTotal = contractLine.getExTaxTotal();
+      int durationType = duration.getTypeSelect();
+      Integer durationValue = duration.getValue();
+      if (initialUnitPrice != null && qty != null && durationValue != null) {
+        contractLine.setInitialPricePerYear(
+            initialUnitPrice
+                .multiply(qty)
+                .multiply(BigDecimal.valueOf(getFactor(durationType) / durationValue.doubleValue()))
+                .setScale(AppBaseService.DEFAULT_NB_DECIMAL_DIGITS, RoundingMode.HALF_UP));
+      }
+      if (exTaxTotal != null && durationValue != null) {
+        contractLine.setYearlyPriceRevalued(
+            exTaxTotal
+                .multiply(BigDecimal.valueOf(getFactor(durationType) / durationValue.doubleValue()))
```

Comment:
same issue as above, also we could use a protected method to avoid duplication of this line